### ISSUE TITLE
Old `thread` analysis is unsound w.r.t. joins

### ISSUE THREAD

**Issue by sim642** [goblint/analyzer#1223]
The old thread analysis is unsound w.r.t. to joins as evidenced by the following program which is incorrectly reported as not racy.

~~~C
// PARAM: --set ana.activated[+] thread
#include <pthread.h>

int data = 0;
pthread_mutex_t data_mutex;

void *thread(void *arg) {
  pthread_mutex_lock(&data_mutex);
  data = 3; // RACE!
  pthread_mutex_unlock(&data_mutex);
  return NULL;
}

int main() {
  pthread_t tids[2];

  pthread_create(&tids[0], NULL, &thread, NULL);
  pthread_create(&tids[1], NULL, &thread, NULL);

  pthread_join(tids[0], NULL);

  data = 1; //RACE!

  return 1;
}
~~~

-------------------------

**Old description for reference:**

Running Goblint on `thread-join-array-const-race` from https://github.com/goblint/bench/pull/63 reveals that our unrolling is unsound w.r.t. thread joins. The program is
```c
// Thread pool constant array sequential joining.
// Extracted from concrat/level-ip.
#include <pthread.h>
extern int __VERIFIER_nondet_int();

int data = 0;
pthread_mutex_t data_mutex = PTHREAD_MUTEX_INITIALIZER;

void *thread(void *arg) {
  pthread_mutex_lock(&data_mutex);
  data = __VERIFIER_nondet_int(); // RACE!
  pthread_mutex_unlock(&data_mutex);
  return NULL;
}

int main() {
  int threads_total = 4;
  pthread_t tids[4];

  // create threads
  for (int i = 0; i < threads_total; i++) {
    pthread_create(&tids[i], NULL, &thread, NULL); // may fail but doesn't matter
  }

  // join threads
  for (int i = 0; i < threads_total - 1; i++) {
    pthread_join(tids[i], NULL);
  }

  return data; // RACE! (tids[3] not joined)
}
```
With the `svcomp` conf, we report no race, which is unsound. If `loopUnrollHeuristic` is disabled, then we become sound again. Therefore, thread joining is unsound when combined with unrolling.

**Comment by michael-schwarz** [comment]
I edited the description above to reflect it is unrelated to loop unrolling.

**Comment by sim642** [comment]
I'm quite surprised this never showed up before. We've used the old thread analysis in SV-COMP for years and with loop unrolling there are a lot more unique thread IDs in SV-COMP tasks. Didn't expect it to be so simple.

I think we've discussed removing the old thread analysis at some point but I no longer recall the details. I suspect we didn't because it can in some case do something threadJoins cannot.

**Comment by michael-schwarz** [comment]
> I suspect we didn't because it can in some case do something threadJoins cannot.

The question is whether this is still the case after we have fixed this issue?
